protocols/ping/src/listen_handler.rs

```diff
@@ -0,0 +1,142 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use arrayvec::ArrayVec;
+use futures::prelude::*;
+use libp2p_core::{
+    nodes::{NodeHandlerEndpoint, ProtocolsHandler, ProtocolsHandlerEvent},
+    ConnectionUpgrade,
+};
+use protocol::{Ping, PingListener, PingOutput};
+use std::io;
+use tokio_io::{AsyncRead, AsyncWrite};
+use void::Void;
+
+/// Handler for handling pings received from a remote.
+pub struct PingListenHandler<TSubstream> {
+    /// Configuration for the ping protocol.
+    ping_config: Ping<()>,
+
+    /// The ping substreams that were opened by the remote.
+    /// Note that we only accept a certain number of substreams, after which we refuse new ones
+    /// to avoid being DDoSed.
+    ping_in_substreams: ArrayVec<[PingListener<TSubstream>; 8]>,
```

Comment:
Is this the right place to mitigate DDOS attacks? Silently ignoring the > 8th pings may cause disruptions if the remotes consider this node unresponsive.

Reply:
Actually, the ping handler is interacting with only a single node. So it is impossible to handle a DDOS attack in here. As far as a single malicious node is concerned, would the stream muxer not be the best place to limit the number of simultaneous streams rather than having each `ProtocolsHandler` be concerned with DOS attacks?

Reply:
I don't have a super strong opinion on that, but the problem to me is that differentiating between a DoS attack and a legitimate use can be a bit tricky on the muxer level.

Reply:
Yes, but the muxer probably limits the number of streams per connection anyway on a level the system can handle.